soroban-env-host/src/host/data_helper.rs

```diff
@@ -122,15 +123,25 @@ impl Host {
         )
     }
 
-    pub(crate) fn retrieve_wasm_from_storage(&self, wasm_hash: &Hash) -> Result<BytesM, HostError> {
+    pub(crate) fn retrieve_wasm_from_storage(
+        &self,
+        wasm_hash: &Hash,
+    ) -> Result<(BytesM, Option<ContractCodeCostInputs>), HostError> {
         let key = self.contract_code_ledger_key(wasm_hash)?;
         match &self
             .try_borrow_storage_mut()?
             .get(&key, self.as_budget())
             .map_err(|e| self.decorate_contract_code_storage_error(e, wasm_hash))?
             .data
         {
-            LedgerEntryData::ContractCode(e) => e.code.metered_clone(self),
+            LedgerEntryData::ContractCode(e) => {
+                let code = e.code.metered_clone(self)?;
+                let costs = match &e.ext {
+                    ContractCodeEntryExt::V0 => None,
```

Comment:
> So old contracts that gets re-uploaded will automatically have such information.

Wasms can not currently be reuploaded and are keyed by the hash of the uploaded Wasm. So if we do nothing, that contracts uploaded during p20 (and maybe even p21, depending on when we roll this out) will never get the extension, unless we do something. I see a few options:

- Do nothing. If most of the contracts are upgradeable, then the owners could upload the new Wasm (maybe with minor no-op changes to change the hash) and get it cheaper. I'm not a big fan of this option because rent for the Wasm upload is pretty expensive and it's not very nice to make the users to either pay it again, or pay more for every invocation.
- Change upload host function to generate the extension even if Wasm is already present on-chain (currently this is a no-op). This seems pretty nice fee-wise, but would require contract maintainers to manually run this operation, which might be not obvious for everyone.
- Lazily generate the extension on Wasm access. This removes the need for anyone to act explicitly, but will introduce unexpected increased fees for the 'unlucky' users that trigger this code path.

Out of these options I would mildly prefer the second one. But maybe there are more options that I'm missing, which is why I'm curious about others' opinions.

Reply:
Ah I see. In this case I think both 2 and 3 would work. 
3 is nicer from the operations perspective. The extension gets populated lazily on invocation, so any contract that gets used becomes cheap to invoke. Although not sure if there is an incentive problem as you mentioned. Maybe the extra cpu isn't as bad in terms of fees. 
2 is good from incentives perspective, shifting the cost from the contract user to the contract code owner/maintainer (who has the incentive to make sure their contracts are cheap to invoke).

Reply:
Actually 3 is kind of risky, at it requires adding a potentially large ledger entry to the RW footprint. Besides the fact that the cost would go up significantly, the transaction might simply exceed the network limit on the write bytes. In the worst case the contract will become completely unusable until it has been updated. So probably we can/should only do 2, even thought it doesn't guarantee full migration.

Reply:
>  Besides the fact that the cost would go up significantly, the transaction might simply exceed the network limit on the write bytes.

Won't this be an issue either way?

Reply:
No, if we only keep 'on-demand' extension creation, then it will be just a single operation that consumes roughly the same resources as a regular Wasm upload. But in the 'lazy' case we would add a non-trivial write that is not generally expected by the dapp.

Reply:
Implemented the "reupload to update it" change.